File: scorecard.py
class Scorecard:
    def __init__(self, num_players):
        
        self.num_players = num_players
        self.cards = [[0] * 13 for _ in range(num_players)]
    
    def get_score(self, player_idx, slot_idx):
        if 0 <= player_idx < self.num_players and 0 <= slot_idx < 13:
            return self.cards[player_idx][slot_idx]
        raise IndexError("Invalid player or slot index")
    
    def set_score(self, player_idx, slot_idx, value):
        if 0 <= player_idx < self.num_players and 0 <= slot_idx < 13:
            self.cards[player_idx][slot_idx] = value
        else:
            raise IndexError("Invalid player or slot index")
    
    def get_player_card(self, player_idx):
        if 0 <= player_idx < self.num_players:
            return self.cards[player_idx]
        raise IndexError("Invalid player index")

File: yahtzee_scala/src/main/scala/yahtzee/Scorecard.scala
package yahtzee

case class Scorecard(numPlayers: Int, cards: Map[(Int, Int), Int]):
  require(numPlayers > 0, "Must have at least one player")
  
  def getScore(playerIdx: Int, slotIdx: Int): Either[String, Int] =
    if playerIdx < 0 || playerIdx >= numPlayers then 
      Left("Invalid player index")
    else if slotIdx < 0 || slotIdx >= 13 then 
      Left("Invalid slot index")
    else 
      Right(cards.getOrElse((playerIdx, slotIdx), 0))
  
  def setScore(playerIdx: Int, slotIdx: Int, value: Int): Either[String, Scorecard] =
    if playerIdx < 0 || playerIdx >= numPlayers then 
      Left("Invalid player index")
    else if slotIdx < 0 || slotIdx >= 13 then 
      Left("Invalid slot index")
    else 
      Right(copy(cards = cards + ((playerIdx, slotIdx) -> value)))
  
  def getPlayerCard(playerIdx: Int): Either[String, List[Int]] =
    if playerIdx < 0 || playerIdx >= numPlayers then 
      Left("Invalid player index")
    else 
      Right((0 until 13).map(slot => cards.getOrElse((playerIdx, slot), 0)).toList)

object Scorecard:
  def apply(numPlayers: Int): Scorecard = 
    Scorecard(numPlayers, Map.empty)
